File: sdk/python/legend_ai/resources/patterns.py
"""
Pattern detection resource
"""

from typing import Optional, TYPE_CHECKING
from ..models import PatternResult

if TYPE_CHECKING:
    from ..client import LegendAI, AsyncLegendAI


class PatternsResource:
    """Synchronous pattern detection resource"""

    def __init__(self, client: "LegendAI"):
        self._client = client

    def detect(
        self,
        ticker: str,
        interval: str = "1day",
        use_yahoo_fallback: bool = False,
    ) -> PatternResult:
        """
        Detect chart patterns for a ticker

        Args:
            ticker: Stock ticker symbol (e.g., "AAPL")
            interval: Time interval ("1day", "1week", "1hour", "4hour")
            use_yahoo_fallback: Use Yahoo Finance as fallback

        Returns:
            PatternResult with detected pattern and scores

        Example:
            >>> result = client.patterns.detect("AAPL")
            >>> print(f"{result.pattern}: {result.score}/10")
        """
        response = self._client.request(
            "POST",
            "/api/patterns/detect",
            json={
                "ticker": ticker,
                "interval": interval,
                "use_yahoo_fallback": use_yahoo_fallback,
            },
        )
        return PatternResult.from_dict(response)

    def health(self) -> dict:
        """Check pattern service health"""
        return self._client.request("GET", "/api/patterns/health")


class AsyncPatternsResource:
    """Asynchronous pattern detection resource"""

    def __init__(self, client: "AsyncLegendAI"):
        self._client = client

    async def detect(
        self,
        ticker: str,
        interval: str = "1day",
        use_yahoo_fallback: bool = False,
    ) -> PatternResult:
        """
        Detect chart patterns for a ticker (async)

        Args:
            ticker: Stock ticker symbol (e.g., "AAPL")
            interval: Time interval ("1day", "1week", "1hour", "4hour")
            use_yahoo_fallback: Use Yahoo Finance as fallback

        Returns:
            PatternResult with detected pattern and scores

        Example:
            >>> result = await client.patterns.detect("AAPL")
            >>> print(f"{result.pattern}: {result.score}/10")
        """
        response = await self._client.request(
            "POST",
            "/api/patterns/detect",
            json={
                "ticker": ticker,
                "interval": interval,
                "use_yahoo_fallback": use_yahoo_fallback,
            },
        )
        return PatternResult.from_dict(response)

    async def health(self) -> dict:
        """Check pattern service health"""
        return await self._client.request("GET", "/api/patterns/health")

File: sdk/javascript/src/resources/patterns.ts
/**
 * Pattern detection resource
 */

import type { AxiosInstance } from 'axios';
import type { PatternDetectRequest, PatternDetectResponse, PatternResult } from '../types';

export class PatternsResource {
  constructor(private client: AxiosInstance) {}

  /**
   * Detect chart patterns for a ticker
   *
   * @param ticker - Stock ticker symbol (e.g., "AAPL")
   * @param options - Detection options
   * @returns Pattern detection result
   *
   * @example
   * ```typescript
   * const pattern = await client.patterns.detect('AAPL', {
   *   interval: '1day'
   * });
   * console.log(`Pattern: ${pattern.pattern}, Score: ${pattern.score}`);
   * ```
   */
  async detect(
    ticker: string,
    options: Omit<PatternDetectRequest, 'ticker'> = {}
  ): Promise<PatternResult> {
    const response = await this.client.post<PatternDetectResponse>('/api/patterns/detect', {
      ticker,
      interval: options.interval || '1day',
      use_yahoo_fallback: options.use_yahoo_fallback || false,
    });
    return response.data.data;
  }

  /**
   * Check pattern service health
   */
  async health(): Promise<Record<string, any>> {
    const response = await this.client.get('/api/patterns/health');
    return response.data;
  }
}
